src/render.rs

```diff
@@ -345,6 +364,81 @@ impl<'a> Renderer<'a> {
         Ok(output.trim_right().to_string())
     }
 
+    fn render_macro(&mut self, call_node: Node) -> TeraResult<String> {
+        if let MacroCall {namespace, name: macro_name, params: call_params} = call_node {
+            // We need to find the active namespace in Tera if `self` is used
+            // Since each macro (other than the `self` ones) pushes its own namespace
+            // to the stack when being rendered, we can just lookup the last namespace that was pushed
+            // to find out the active one
+            let active_namespace = match namespace.as_ref() {
+                "self" => {
+                    // TODO: handle error if we don't have a namespace
+                    // This can only happen when calling {{ self:: }} outside of a macro afaik
+                    self.macro_namespaces
+                        .last()
+                        .expect("Open an issue with a template sample please (mention `self namespace macro`)!")
+                        .to_string()
+                },
+                _ => {
+                    // TODO: String doesn't have Copy trait, can we avoid double cloning?
+                    self.macro_namespaces.push(namespace.clone());
+                    namespace.clone()
+                }
+            };
+
+            // We get our macro definition using the namespace name we just got
+            let macro_definition = self.macros
+                .get(&active_namespace)
+                .and_then(|m| m.get(&macro_name).cloned());
+
+            if let Some(Macro {body, params, ..}) = macro_definition {
+                let expected_params = params.iter().cloned().collect::<Vec<String>>();
+                let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
```

Comment:
I can remove them from there but then if i only collect the args on errors, I run into that error:

```rust
                for (param_name, exp) in call_params.clone() {
                    if !params.contains(&param_name) {
                        let expected_params = params.iter().cloned().collect::<Vec<String>>();
                        let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
                        return Err(MacroCallWrongArgs(macro_name, expected_params, params_seen));
                    }
                    context.insert(param_name.to_string(), try!(self.eval_expression(exp)));
                }
```
Is there a way to iterate on a `HashMap` without moving the value and without `clone()`? I need the args name for the error message